Gutenberg: modify BrowserURL to match Calypso routes

gwwar opened the pull request:
This fixes #27298 where the browser url was incorrectly updated to match wp-admin routes vs Calypso ones. There is a separate issue to update the document title appropriately in https://github.com/Automattic/wp-calypso/issues/27300

### Testing Instructions
- Visit http://calypso.localhost:3000/gutenberg
- Select a site
- in console type `state.ui.route.path` verify that this matches the current path
- Type and let an autosave occur

The browser address is updated to:
http://calypso.localhost:3000/gutenberg/post/{siteSlug}/{postId}

Instead of http://calypso.localhost:3000/gutenberg/post/post.php?post={postId}&action=edit

After the url updates, verify that `state.ui.route.path` in console matches

dmsnell commented on client/gutenberg/editor/browser-url/index.js:
```diff
@@ -0,0 +1,63 @@
+
+		this.setState( () => ( {
+			historyId: postId,
+		} ) );
```
we may want to rearrange this here so that the state updates first. I don't know if we could envision a re-render on `replaceState` or not but if that's possible but there's also an almost-circular thing going on here

what if we flipped it so `setState` were in `componentDidUpdate` and the `setBrowserURL` method _only_ set the browser URL?

```js
if ( … ) {
  this.setState( { historyId: postId }, () => this.setBrowserURL( postId ) );
}
```

also was the function syntax in `setState` habit or leftover? since we're assigning a raw value there I don't think there's any difference in the runtime

gwwar replied:
I basically deleted code 🔪 from the base Gutenberg component.

The guts of the update cycle are from the `edit-post` BrowserURL component, but I suppose I could write something smaller from scratch :)

Last 30 minutes of the day trying to sneak in a PR

gwwar replied:
Oh hmm, we have `react/no-did-update-set-state` as an eslint rule. Let me 🤔 a bit more on this.

dmsnell commented on client/gutenberg/editor/browser-url/index.js:
```diff
@@ -0,0 +1,63 @@
+		window.history.replaceState(
+			{ id: postId },
+			null,
+			`${ window.location.href }/${ postId }`
```
we probably need to use the URL-building functions here. `href` could also include a fragment or query string

Copons commented on client/gutenberg/editor/browser-url/index.js:
```diff
@@ -0,0 +1,63 @@
+		this.state = {
+			historyId: null,
+		};
+	}
```
Is this format to keep the file as similar as possible to the Guten version?
Couldn't we move the state as a class property?

gwwar replied:
Copy pasta removed in 41d9a0e

Copons commented:
Tested and it works fine! ✨ 

I've left a minor nitpick and I agree with @dmsnell re: https://github.com/Automattic/wp-calypso/pull/27301#discussion_r218662006, but this looks pretty much ready!

gwwar commented:
Thanks for the reviews! I'll give this a better rewrite today or tomorrow, there isn't much here we need to save from the core component.

kwight commented:
Works great for me too as-is 👍

gwwar commented:
Thanks @dmsnell @Copons @kwight! I think I addressed most feedback in 41d9a0e, though let me know if we think it's overkill to connect to global state for just the route path. I can use `url.parse` otherwise.

dmsnell commented on client/gutenberg/editor/browser-url/index.js:
```diff
@@ -0,0 +1,68 @@
+ * For example:
+ * https://wordpress.com/gutenberg/post/mysiteslug
+ *
+ * https://wordpress.com/gutenberg/post/mysiteslug/1234
```
thanks for this excellent example - so clear

gwwar commented:
I've updated this to use navigation middleware since I realized global routes weren't updated. I'll land this tomorrow unless folks have issues with the approach.

gwwar commented:
Thanks for the reviews folks! 💖
